ir/src/tests/list_folding.rs

```diff
@@ -0,0 +1,107 @@
+use super::{parse, AirIR};
+
+#[test]
+fn list_folding_on_const() {
+    let source = "
+    const A = [1, 2, 3]
+    trace_columns:
+        main: [clk, fmp[2], ctx]
+        aux: [a, b, c[4], d[4]]
+    public_inputs:
+        stack_inputs: [16]
+    boundary_constraints:
+        enf c[2].first = 0
+    integrity_constraints:
+        let x = sum(A)
+        let y = prod(A)
+        enf clk = y - x";
+
+    let parsed = parse(source).expect("Parsing failed");
+    let result = AirIR::new(&parsed);
+    assert!(result.is_ok());
+}
+
+#[test]
+fn list_folding_on_variable() {
+    let source = "
+    trace_columns:
+        main: [clk, fmp[2], ctx]
+        aux: [a, b, c[4], d[4]]
+    public_inputs:
+        stack_inputs: [16]
+    boundary_constraints:
+        enf c[2].first = 0
+    integrity_constraints:
+        let x = [a + c, 1, c[2] * d[2]]
+        let y = sum(x)
+        let z = prod(x)
+        enf clk = z - y";
+
+    let parsed = parse(source).expect("Parsing failed");
+    let result = AirIR::new(&parsed);
+    assert!(result.is_ok());
```

Comment:
Do we allow the syntax `a+c`? I think we probably shouldn't - it should only be `a+c[i]` that's supported. Since this is passing, maybe there's a bug to fix (and a corresponding new test to add).

Reply:
Thanks for catching this. I've included a fix in this PR.